palette/src/rgb/packed.rs

```diff
@@ -0,0 +1,364 @@
+use core::marker::PhantomData;
+
+use crate::rgb::{Rgb, RgbStandard, Rgba};
+use crate::{Alpha, Pixel};
+
+/// RGBA color packed into a 32-bit unsigned integer, defaults to ARGB ordering.
+///
+/// Packed integer type represented in `u32`. Two hexadecimal digits (8-bits)
+/// express each value of the Red, Green, Blue, and Alpha components in the
+/// RGBA color.
+///
+/// Note that conversion from float to integer component types in `palette` maps
+/// the floating point value to the integer's value range and then rounds the
+/// result before casting to the integer type. An example of the difference in
+/// conversion methods is shown below.
+///
+/// ```
+/// use approx::assert_relative_eq;
+/// use palette::{PackedRgb, Srgb, Srgba};
+/// use palette::rgb::{Argb, RgbaPacked};
+///
+/// let packed: PackedRgb = Srgb::new(0.5, 0.0, 0.5).into_format().into();
+/// assert_eq!(0x0080_0080, packed.color);
+///
+/// let unpacked: Srgba<u8> = PackedRgb::<RgbaPacked>::from(0x80FF_FF80).into();
+/// assert_relative_eq!(
+///     Srgba::new(0.5, 1.0, 1.0, 0.5),
+///     unpacked.into_format(),
+///     epsilon = 0.01
+/// );
+///
+/// // By default, `PackedRGB` uses `Argb` order to make entering 6-digit hex
+/// // numbers more convenient
+/// let unpacked = Srgb::<u8>::from_hex::<Argb>(0xFF8000);
+/// assert_eq!(Srgb::<u8>::new(0xFF, 0x80, 0x00), unpacked);
+///
+/// // The second assert is essentially how library components are converted
+/// let float = 0.5f32 * 255.0;
+/// assert_eq!(0x7F, float as u32);
+/// assert_eq!(0x80, float.round() as u32);
+/// ```
+///
+/// When an `Srgb` type is packed, the alpha value will be `0x00` in the
+/// corresponding `u32`. Converting from a packed color type back to an `Srgb`
+/// type will disregard the alpha value.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Pixel)]
+#[palette_internal]
+#[repr(C)]
+pub struct PackedRgb<C: RgbChannels = Argb> {
+    /// The sRGB color packed into a `u32`.
+    pub color: u32,
+
+    /// The channel ordering for red, green, blue, and alpha components in the
+    /// packed integer; can be `Argb`, `Abgr`, `Bgra`, or `RgbaPacked`.
+    #[palette_unsafe_zero_sized]
+    pub channel_order: PhantomData<C>,
+}
+
+/// Splits and combines RGB(A) types with some channel ordering. Channels may be
+/// ordered as `Argb`, `Abgr`, `Bgra`, or `RgbaPacked`.
+pub trait RgbChannels {
+    /// Split RGBA components into a `(u8, u8, u8, u8)` tuple.
+    fn split_rgb<S: RgbStandard>(rgb: Rgba<S, u8>) -> (u8, u8, u8, u8);
+    /// Create an RGBA color from a `(u8, u8, u8, u8)` tuple.
+    fn combine_rgb<S: RgbStandard>(channels: (u8, u8, u8, u8)) -> Rgba<S, u8>;
+}
+
+/// RGBA color packed in RGBA order.
+///
+/// See [PackedRgb](struct.PackedRgb.html) for more details.
+pub struct RgbaPacked;
```

Comment:
The fact that this is in a separate module makes it possible to tell them apart, so I think the `Packed` part can be dropped. It's still annoying that color spaces are also named after their channels. It could perhaps be further emphasizes with a module for these types (like `channels::Rgba`), or by going a bit unorthodox and using all capitals. Using all caps may be confusing, though.